downloader/src/main/kotlin/VcsHost.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2019 Bosch Software Innovations GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.downloader
+
+import com.here.ort.model.VcsInfo
+import com.here.ort.model.VcsType
+import com.here.ort.utils.hasFragmentRevision
+
+import java.net.URI
+import java.net.URISyntaxException
+import java.nio.file.Paths
+
+import kotlin.reflect.full.primaryConstructor
+
+/**
+ * A class to handle VCS-host-specific information.
+ */
+sealed class VcsHost(val url: String) {
+    companion object {
+        private val ALL = VcsHost::class.sealedSubclasses
+
+        /**
+         * Get the VCS host that is applicable for [url].
+         */
+        fun fromUrl(url: String): VcsHost? =
+            ALL.asSequence().map {
+                it.primaryConstructor!!.call(url)
+            }.find {
+                it.isApplicable()
+            }
+    }
+
+    /**
+     * The [URI] for [url] if it can be parsed, or null otherwise.
+     */
+    protected val uri = try {
+        URI(url)
+    } catch (e: URISyntaxException) {
+        null
+    }
+
+    /**
+     * The base hostname of VCS host.
+     */
+    abstract val hostname: String
+
+    /**
+     * Whether the VCS host can only host Git repositories.
+     */
+    abstract val isGitOnly: Boolean
+
+    /**
+     * Return whether this instance is applicable for the [uri] it was constructed with.
+     */
+    fun isApplicable() = uri?.host == hostname
+
+    /**
+     * Return all [VcsInfo] that can be extracted from [uri].
+     */
+    abstract fun toVcsInfo(): VcsInfo?
+}
+
+/**
+ * A class to handle [Bitbucket][https://bitbucket.org/]-specific information.
+ */
+class Bitbucket(url: String) : VcsHost(url) {
+    override val hostname = "bitbucket.org"
+    override val isGitOnly = false
+
+    override fun toVcsInfo(): VcsInfo? {
```

Comment:
Why did you decide to create one instance per URL, e.g. the Url is passed to the constructor and this function does not take any parameter. Extracting a VcsInfo out of a URL to me seems more like a utility function not requiring one instance per parameter value.

Reply:
The reason for that design decision was to be able to write code like `VcsHost.fromUrl(homepage)?.toVcsInfo()`. If a VcsHost instance would not know about its URL, the URL would need to be passed to `toVcsInfo()` in the preceding example, making the code look like `VcsHost.fromUrl(homepage)?.toVcsInfo(homepage)`, which I found to be very ugly because of the repetition of the `homepage` argument.

Reply:
It becomes clearer to me but I still need to ask one more thing. As I understand it the responsibility of this VcsHost class is to transform URLs to VcsInfo. Am I right that the caller would not need an instance for this at all?

Subclassing and instance creation could be completely made irrelevant to the API user, e.g. by allowing to call ```VcsHost.urlToVcsInfo(homepage)```. How would you compare that one with the current ```VcsHost.fromUrl(homepage)?.toVcsInfo()```?

Reply:
> As I understand it the responsibility of this VcsHost class is to transform URLs to VcsInfo.

That's the only thing it *currently* does, yes, but I plan to add more host-specific functionality soon, like converting the other way around, `VcsInfo` to browsable URLs, and maybe even some code that talks to the respective REST APIs.

> VcsHost.urlToVcsInfo(homepage)

That call internally then would have to implement the same logic as `fromUrl(homepage)` for the host-specific extraction of VCS info. As soon as another similar function would get implemented, the logic would be needed again, so you'd probably start extracting it to a helper function, which is what `fromUrl(homepage)` basically is.

Reply:
Now that I'm aware of these future plans to extend the code it makes sense.
Let's just keep as is, it wouldn't be expensive to change anyway in case it turned out to make sense later on.

Reply:
> Why did you decide to create one instance per URL

FYI, I indeed changed that now in https://github.com/heremaps/oss-review-toolkit/pull/1881/commits/afb4063d8960904d4842b37ecec7436beac150a6.